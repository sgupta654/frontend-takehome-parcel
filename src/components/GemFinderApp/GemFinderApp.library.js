import { css } from '@emotion/core';
import ls from 'local-storage';

/*
  Gets saved gems from localStorage
*/
export const getLSGems = () => {
  let lsCurrent = ls.get('savedGems');

  if(lsCurrent === null || lsCurrent === undefined) {
    return [];
  }
  else {
    return JSON.parse(lsCurrent);
  }
}

export const savedStyles = (viewSavedGems) => css`
  opacity: 0;
  ${viewSavedGems && `
    transition: opacity 1s ease-out;
    opacity: 1;
    overflow: hidden;
  `};
  vertical-align: top;
  min-width: 50%;
  float: right;
`;

export const savedHeaderStyles = css`
  text-align: center;
  line-height: 2.5rem;
  color: hsl(163, 72%, 47%);
  font-size: 1.5rem;
`;

export const searchResultsStyles = css`
  min-width: 50%;
  float: left;
`;

export const searchResultsHeaderStyles = css`
  text-align: center;
  line-height: 2.5rem;
  color: hsl(163, 72%, 47%);
  font-size: 1.5rem;
`;

export const appHeaderStyles = css`
  background-color: hsl(163, 72%, 47%);
  min-height: 5rem;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

export const savedButtonStyles = css`
  margin-right: 1rem;
  min-height: 2.4rem;
  font-size: 24px;
  color: hsl(163, 72%, 47%);
  background-color: hsl(0, 0%, 100%);
  border-radius: .5rem;
  border: .2rem solid hsl(0, 0%, 100%);
  padding: .6rem;
  &:active {
    border: .2rem solid hsl(0, 0%, 100%);
    background-color: hsl(163, 72%, 47%);
    color: hsl(0, 0%, 100%);
  }
`;




