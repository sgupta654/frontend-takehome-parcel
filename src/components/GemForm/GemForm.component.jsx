/** @jsx jsx */

import { Button } from "../Button/Button.component";
import PropTypes from 'prop-types';
import { jsx } from '@emotion/core';
import { gemFormButtonStyles, gemFormInputStyles } from "./GemForm.library";

export const GemForm = (props) => {

  const {className, onChange, onSubmit, query} = props;

  const onSubmitHandler = () => {
    event.preventDefault();
    onSubmit(query);
  };

  const onChangeHandler = (e) => {
    onChange(e.target.value);
  }

  return (
    <form onSubmit={onSubmitHandler} css={className}>
      <input
        css={gemFormInputStyles}
        placeholder={"Search for a gem..."} 
        type="text" 
        onChange={onChangeHandler} 
        value={query} 
      />
      <input css={gemFormButtonStyles}
        type="submit" 
        value="Search Gems"
      />
    </form>
  );
};

Button.propTypes = {
  className: PropTypes.string,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  query: PropTypes.string
};