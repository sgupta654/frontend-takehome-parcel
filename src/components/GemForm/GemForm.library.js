import { css } from '@emotion/core';

export const gemFormInputStyles = css`
  border: 0.2rem solid transparent;
  border-radius: 1rem;
  font-size: 1.5rem;
  margin-right: 1rem;
  min-height: 2rem;
  padding: 0.5rem;

  &:focus {
    border: 0.2rem solid hsl(0, 0%, 0%);
    box-shadow: 0 0 .2rem hsl(216, 100%, 50%);
    outline: none;
  }
`;

export const gemFormButtonStyles = css`
  background-color: hsl(0, 0%, 100%);
  border: .2rem solid hsl(0, 0%, 100%);
  border-radius: 0.5rem;
  color: hsl(163, 72%, 47%);
  font-size: 1.5rem;
  min-height: 2.4rem;
  padding: 0.6rem;

  &:active {
    border: 0.2rem solid hsl(0, 0%, 100%);
    background-color: hsl(163, 72%, 47%);
    color: hsl(0, 0%, 100%);
  }
`;