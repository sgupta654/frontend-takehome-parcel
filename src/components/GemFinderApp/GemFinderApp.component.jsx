/** @jsx jsx */

import { Button } from "../Button/Button.component";
import { getLSGems } from "./GemFinderApp.library";
import { Gem } from "../Gem/Gem.component";
import { GemForm } from "../GemForm/GemForm.component";
import { List } from "../List/List.component";
import ls from 'local-storage';
import React from "react";
import { sendAPIQuery } from "../../utils/api/api";

import { css, jsx } from '@emotion/core';
import { useCallback, useEffect, useState } from "react";

import { appHeaderStyles,
         savedStyles, 
         savedHeaderStyles,
         savedButtonStyles, 
         searchResultsStyles,
         searchResultsHeaderStyles } from "./GemFinderApp.library";

const GemFinderApp = () => {
  const [userQuery, setUserQuery] = useState("");
  const [searchedGems, setSearchedGems] = useState([]);
  const [savedGems, setSavedGems] = useState(getLSGems);
  const [viewSavedGems, setViewSavedGems] = useState(false);

  const addSavedGem = useCallback(
    (gem) => {
      let newSavedGems;
      setSavedGems(oldGems => {
        if(oldGems.findIndex(g => g.name === gem.name) === -1) {
          newSavedGems = [...oldGems, gem];
          ls.set("savedGems", JSON.stringify(newSavedGems));
          return [...oldGems, gem];
        } else {
          alert(gem.name + " already saved!");
          return oldGems;
        }
      });
    },
    [setSavedGems],
  );

  const removeSavedGem = useCallback(
    (gem) => {
      setSavedGems(oldGems => {
        let newGemsSaved = oldGems.filter(g => g.name !== gem.name);
        ls.set("savedGems", JSON.stringify(newGemsSaved));
        return newGemsSaved;
      });
    },
    [setSavedGems],
  )

  const parseGems = useCallback(
    (gems) => {
      let parsedGems = gems.map((gem) => {
        return {
          id: gem.name,
          name: gem.name,
          version: gem.version,
          saved: savedGems.some(g => g.name === gem.name)
        }
      });

      setSearchedGems(parsedGems);
    },
    [setSearchedGems],
  );

  const sendQuery = useCallback(async () => {
    const apiGems = await sendAPIQuery("search.json?query=", userQuery);
    parseGems(apiGems);
  },
  [userQuery]);


  const renderSearchGemOption = useCallback((gemProps) => {
    return(
      <Gem key={gemProps.id} onClick={addSavedGem} {...gemProps} />
    )
  });

  const renderGemOption = useCallback((gemProps) => {
    return(
      <Gem key={gemProps.id} onClick={removeSavedGem} {...gemProps} />
    )
  });

  return (
    <>
      <div css={appHeaderStyles}>
        <GemForm css={css`
            margin-left: 1rem;
          `} 
          onChange={setUserQuery}
          onSubmit={sendQuery} 
          query={userQuery} 
        />
        <Button css={savedButtonStyles}
        onClick={() => {setViewSavedGems(true)}}
        >View Saved Gems</Button>
      </div>
      <div css={searchResultsStyles}>
        <div css={searchResultsHeaderStyles}>Search Results: {searchedGems.length}</div>
        <List id="searchedGemsList" renderOption={renderSearchGemOption} options={searchedGems} />
      </div>
      <div css={savedStyles(viewSavedGems)}>
        <div css={savedHeaderStyles}>Saved Gems: {savedGems.length}</div>
        <List id="savedGemsList" renderOption={renderGemOption} options={savedGems} />
      </div>
    </>
  )
};

export default GemFinderApp;