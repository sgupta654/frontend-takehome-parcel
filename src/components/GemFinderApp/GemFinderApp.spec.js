import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme'

import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

import GemFinderApp from "./GemFinderApp.component";
import { GemForm } from '../GemForm/GemForm.component';

describe('GemFinderApp', () => {
  test('snapshot renders', () => {
    const component = renderer.create(<GemFinderApp />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders without crashing", () => {
    shallow(<GemFinderApp />);
  });
});