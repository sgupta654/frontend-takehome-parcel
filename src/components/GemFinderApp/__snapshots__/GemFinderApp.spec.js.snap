// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GemFinderApp snapshot renders 1`] = `
Array [
  <div
    className="css-1dx9on5-GemFinderApp"
  >
    <form
      className="css-d2rifm-GemFinderApp"
      onSubmit={[Function]}
    >
      <input
        className="css-1pdxv3p-gemFormInputStyles"
        onChange={[Function]}
        placeholder="Search for a gem..."
        type="text"
        value=""
      />
      <input
        className="css-spbcps-gemFormButtonStyles"
        type="submit"
        value="Search Gems"
      />
    </form>
    <button
      className="css-1jpq8qr-GemFinderApp-GemFinderApp"
      onClick={[Function]}
    >
      View Saved Gems
    </button>
  </div>,
  <div
    className="css-874azv-GemFinderApp"
  >
    <div
      className="css-8uqa1p-GemFinderApp"
    >
      Search Results: 
      0
    </div>
    <ul
      className="css-ylle7e-listStyles"
      id="searchedGemsList"
    />
  </div>,
  <div
    className="css-1emp3qb-GemFinderApp"
  >
    <div
      className="css-8uqa1p-GemFinderApp"
    >
      Saved Gems: 
      0
    </div>
    <ul
      className="css-ylle7e-listStyles"
      id="savedGemsList"
    />
  </div>,
]
`;
